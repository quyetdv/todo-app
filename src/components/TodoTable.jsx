import { Table, Button } from 'antd';
import { useContext } from 'react';
import PropTypes from 'prop-types';

import TodoListContext from '../context/TodoList';

const priorityMapper = (data) => {
  const mapper = { 0: 'High', 1: 'Medium', 2: 'Low' };
  return mapper[data];
};

function TodoTable({ onView, onMarkDone, onDelete }) {
  const { todoList } = useContext(TodoListContext);

  const footer = () => {
    return (
      <>
        <span className="footer-span">Total: {todoList.length}</span>
        <span className="footer-span">
          Completed: {todoList.filter((todoItem) => todoItem.completed).length}
        </span>
      </>
    );
  };

  const columns = [
    {
      title: 'Task',
      dataIndex: 'task',
      key: 'task',
      render: (text, record) => (
        <span
          style={{ textDecoration: record.completed ? 'line-through' : null }}
        >
          {text}
        </span>
      ),
      sorter: (a, b) => a.task - b.task,
      width: '30%',
    },
    {
      title: 'Priority',
      dataIndex: 'priority',
      key: 'priority',
      render: (text, record) => (
        <span
          style={{ textDecoration: record.completed ? 'line-through' : null }}
        >
          {priorityMapper(record.priority)}
        </span>
      ),
      sorter: (a, b) => a.priority - b.priority,
      width: '10%',
    },
    {
      title: 'Description',
      dataIndex: 'description',
      key: 'description',
      render: (text, record) => (
        <span
          style={{ textDecoration: record.completed ? 'line-through' : null }}
        >
          {text}
        </span>
      ),
      width: '40%',
    },
    {
      title: 'Action',
      dataIndex: '',
      key: 'x',
      render: (text, record) => (
        <div className="button-group">
          <Button
            onClick={() => {
              onView(record.key);
            }}
          >
            {record.completed ? 'View' : 'Edit'}
          </Button>
          <Button
            // style={{ backgroundColor: 'green', color: 'white' }}
            disabled={record.completed}
            onClick={() => {
              onMarkDone(record.key);
            }}
          >
            Mark As Done
          </Button>
          <Button
            type="danger"
            onClick={() => {
              onDelete(record.key);
            }}
          >
            Delete
          </Button>
        </div>
      ),
      width: '20%',
    },
  ];

  return (
    <Table
      className="todo-table"
      columns={columns}
      dataSource={todoList}
      footer={footer}
      bordered
    />
  );
}

TodoTable.propTypes = {
  onView: PropTypes.func,
  onMarkDone: PropTypes.func,
  onDelete: PropTypes.func,
};

TodoTable.defaultProps = {
  onView: () => {},
  onMarkDone: () => {},
  onDelete: () => {},
};

export default TodoTable;
