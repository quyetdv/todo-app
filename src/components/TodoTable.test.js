import React from 'react';
import renderer from 'react-test-renderer';
import TodoTable from './TodoTable';

describe('TodoTable', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('renders TodoTable component', () => {
    const component = renderer.create(<TodoTable />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
