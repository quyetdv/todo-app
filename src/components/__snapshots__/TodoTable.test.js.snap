// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoTable renders TodoTable component 1`] = `
<div
  className="ant-table-wrapper todo-table"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-table ant-table-bordered"
      >
        <div
          className="ant-table-container"
        >
          <div
            className="ant-table-content"
            onScroll={[Function]}
            style={Object {}}
          >
            <table
              style={
                Object {
                  "tableLayout": "auto",
                }
              }
            >
              <colgroup>
                <col
                  style={
                    Object {
                      "width": "30%",
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "width": "10%",
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "width": "40%",
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "width": "20%",
                    }
                  }
                />
              </colgroup>
              <thead
                className="ant-table-thead"
              >
                <tr>
                  <th
                    className="ant-table-cell ant-table-column-has-sorters"
                    colSpan={null}
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <div
                      className="ant-table-column-sorters"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <span
                        className="ant-table-column-title"
                      >
                        Task
                      </span>
                      <span
                        className="ant-table-column-sorter ant-table-column-sorter-full"
                      >
                        <span
                          className="ant-table-column-sorter-inner"
                        >
                          <span
                            aria-label="caret-up"
                            className="anticon anticon-caret-up ant-table-column-sorter-up"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="caret-up"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="0 0 1024 1024"
                              width="1em"
                            >
                              <path
                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                              />
                            </svg>
                          </span>
                          <span
                            aria-label="caret-down"
                            className="anticon anticon-caret-down ant-table-column-sorter-down"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="caret-down"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="0 0 1024 1024"
                              width="1em"
                            >
                              <path
                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </div>
                  </th>
                  <th
                    className="ant-table-cell ant-table-column-has-sorters"
                    colSpan={null}
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <div
                      className="ant-table-column-sorters"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <span
                        className="ant-table-column-title"
                      >
                        Priority
                      </span>
                      <span
                        className="ant-table-column-sorter ant-table-column-sorter-full"
                      >
                        <span
                          className="ant-table-column-sorter-inner"
                        >
                          <span
                            aria-label="caret-up"
                            className="anticon anticon-caret-up ant-table-column-sorter-up"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="caret-up"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="0 0 1024 1024"
                              width="1em"
                            >
                              <path
                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                              />
                            </svg>
                          </span>
                          <span
                            aria-label="caret-down"
                            className="anticon anticon-caret-down ant-table-column-sorter-down"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="caret-down"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="0 0 1024 1024"
                              width="1em"
                            >
                              <path
                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </div>
                  </th>
                  <th
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    Description
                  </th>
                  <th
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    Action
                  </th>
                </tr>
              </thead>
              <tbody
                className="ant-table-tbody"
              >
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key={0}
                  onClick={[Function]}
                  style={Object {}}
                >
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      style={
                        Object {
                          "textDecoration": null,
                        }
                      }
                    >
                      
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      style={
                        Object {
                          "textDecoration": null,
                        }
                      }
                    >
                      High
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <span
                      style={
                        Object {
                          "textDecoration": null,
                        }
                      }
                    >
                      
                    </span>
                  </td>
                  <td
                    className="ant-table-cell"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={Object {}}
                  >
                    <div
                      className="button-group"
                    >
                      <button
                        className="ant-btn ant-btn-default"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          Edit
                        </span>
                      </button>
                      <button
                        className="ant-btn ant-btn-default"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          Mark As Done
                        </span>
                      </button>
                      <button
                        className="ant-btn ant-btn-danger"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          Delete
                        </span>
                      </button>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          className="ant-table-footer"
        >
          <span
            className="footer-span"
          >
            Total: 
            1
          </span>
          <span
            className="footer-span"
          >
            Completed: 
            0
          </span>
        </div>
      </div>
      <ul
        className="ant-pagination ant-table-pagination ant-table-pagination-right"
        style={Object {}}
        unselectable="unselectable"
      >
        <li
          aria-disabled={true}
          className="ant-pagination-prev ant-pagination-disabled"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex={null}
          title="Previous Page"
        >
          <button
            className="ant-pagination-item-link"
            disabled={true}
            tabIndex={-1}
            type="button"
          >
            <span
              aria-label="left"
              className="anticon anticon-left"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                />
              </svg>
            </span>
          </button>
        </li>
        <li
          className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex="0"
          title={1}
        >
          <a
            rel="nofollow"
          >
            1
          </a>
        </li>
        <li
          aria-disabled={true}
          className="ant-pagination-next ant-pagination-disabled"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex={null}
          title="Next Page"
        >
          <button
            className="ant-pagination-item-link"
            disabled={true}
            tabIndex={-1}
            type="button"
          >
            <span
              aria-label="right"
              className="anticon anticon-right"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                />
              </svg>
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
